table.cc

```diff
@@ -661,7 +661,11 @@ table::try_flush_memtable_to_sstable(lw_shared_ptr<memtable> old, sstable_write_
                         return update_cache(old, newtabs);
                     });
                 }).then([this, old, &newtabs] () noexcept {
-                    _memtables->erase(old);
+                    // # 8749 - if a table::clear() was issued while we were flushing,
+                    // the memtable is already gone. Check before erase.
+                    if (_memtables->contains(old)) {
+                        _memtables->erase(old);
+                    }
```

Comment:
I can commit this, but to me, this shows that our erase() API is bad. It's really weird to have a comment saying that "The caller has to make sure the element exist before calling this." - and then all callers do if(contains) erase(). Would shouldn't erase() just check the result of find() and ignore the erase if find found nothing?
It will also be more efficient than doing the find() twice...

Reply:
I opted to keep the erase as it is, as it mimics the behaviour of stl containers. Alternatively rename erase to "erase_if_exists", _or_ make it throw if it does not. 
But again, opted for stl-behaviour.

Reply:
Btw, credit goes to @bhalevy who finally saw the glaring error. I missed it just because I assumed erase of non-existent would be no-op. I am not sure which approach that is arguing for... Either that stl way is wrong, or that we have too many api:s that are "forgiving" (java-esque). 

One argument for changing "erase" would maybe be that stl variants which require stuff to exist are iterator variants. Here is more set::erase(key), which _is_ forgiving. I think I am changing my mind...

Reply:
Changed erase instead.

Reply:
But @elcallio your statement about STL erase is not true... `std::unordered_map<>::erase(key)` doesn't do anything (and definitely doesn't crash) if key doesn't exist. I see you also added that to your second comment...
It's also a matter of performance - calling contains() and then erase() just runs the find() twice, which is silly.
It's also a matter of the risk to future code that uses this erase() (or even existing code - did you check if it's used anywhwere else)?

Reply:
It is true for vector::erase(iter), which memtable list wraps. But as I wrote earlier, on second consideration, the api here is a set-like one, so I do agree that erase should be tolerant. Already changed.

Reply:
Sorry, my pushing skills failed me. Now it is re-pushed.